r/src/tpch-gen.cpp

```diff
@@ -0,0 +1,169 @@
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include "./arrow_types.h"
+
+#if defined(ARROW_R_WITH_ARROW)
+
+// TODO: prune these
+#include <arrow/compute/api.h>
+#include <arrow/compute/exec/exec_plan.h>
+#include <arrow/compute/exec/tpch_node.h>
+#if defined(ARROW_R_WITH_DATASET)
+#include <arrow/dataset/api.h>
+#include <arrow/dataset/plan.h>
+#include <arrow/filesystem/filesystem.h>
+#include <arrow/filesystem/localfs.h>
+
+namespace compute = ::arrow::compute;
+namespace ds = ::arrow::dataset;
+namespace fs = ::arrow::fs;
+
+static std::shared_ptr<compute::ExecNode> MakeExecNodeOrStop(
+    const std::string& factory_name, compute::ExecPlan* plan,
+    std::vector<compute::ExecNode*> inputs, const compute::ExecNodeOptions& options) {
+  return std::shared_ptr<compute::ExecNode>(
+      ValueOrStop(compute::MakeExecNode(factory_name, plan, std::move(inputs), options)),
+      [](...) {
+        // empty destructor: ExecNode lifetime is managed by an ExecPlan
+      });
+}
+
+// [[arrow::export]]
+std::shared_ptr<arrow::RecordBatchReader> Tpch_Dbgen(
+    const std::shared_ptr<compute::ExecPlan>& plan, double scale_factor,
+    std::string table_name) {
+  auto gen =
+      ValueOrStop(arrow::compute::internal::TpchGen::Make(plan.get(), scale_factor));
+
+  compute::ExecNode* table;
+  if (table_name == "part") {
+    table = ValueOrStop(gen->Part());
+  } else if (table_name == "supplier") {
+    table = ValueOrStop(gen->Supplier());
+  } else if (table_name == "partsupp") {
+    table = ValueOrStop(gen->PartSupp());
+  } else if (table_name == "customer") {
+    table = ValueOrStop(gen->Customer());
+  } else if (table_name == "nation") {
+    table = ValueOrStop(gen->Nation());
+  } else if (table_name == "lineitem") {
+    table = ValueOrStop(gen->Lineitem());
+  } else if (table_name == "region") {
+    table = ValueOrStop(gen->Region());
+  } else if (table_name == "orders") {
+    table = ValueOrStop(gen->Orders());
+  } else {
+    cpp11::stop("That's not a valid table name");
+  }
+
+  arrow::AsyncGenerator<arrow::util::optional<compute::ExecBatch>> sink_gen;
+
+  MakeExecNodeOrStop("sink", plan.get(), {table}, compute::SinkNodeOptions{&sink_gen});
+
+  StopIfNotOk(plan->Validate());
+  StopIfNotOk(plan->StartProducing());
```

Comment:
It looks like you're starting the same plan multiple times, which doesn't seem write. You're starting it here and then at the end in Tpch_Dbgen_Write